### PR TITLE
Bug/randomness

### LINKED ISSUE #798: uncontrolled randomness
The results from one simulation are slightly different if re-computed. Did we miss a seed somewhere? 

@HenryGeorgist if you have a couple minutes, I'd appreciate your help investigating. The means are the same but the distributions are not. The quartiles are off by less than $100 - a very small fraction. I am not sure if it matters, but it looks like the difference is increasing with increasing cumulative probability.  

My greater concern for **this** is that I think the issue is related to #782. 

## Existing Condition Future Year Left Bank EAD Results
image

## Same Results Recomputed 
image

### PR DESCRIPTION
**Pull request by rnugent3**
This Fixes #798.

### CODE REVIEW

File: HEC.FDA.Model/compute/ImpactAreaScenarioSimulation.cs [hunk 1]
```diff
@@ -18,6 +18,7 @@
 using System.Collections.Concurrent;
 using Statistics.Distributions;
 using HEC.FDA.Model.extensions;
+using Microsoft.VisualBasic;
```
Comment (rnugent3) [inline HEC.FDA.Model/compute/ImpactAreaScenarioSimulation.cs:21]: I am not sure why this was added. Is it a problem?

**Comment by rnugent3** [comment]
I think it would be possible to obtain performance improvements (reduction in memory requirements) by chopping up min iterations into 10 or 20 chunks.

**Comment by rnugent3** [comment]
Stage damage tests are failing due to memory issues

**Review by Brennan1994 — COMMENTED** [review]
I might need this explained to me rather than giving it an independent review. I'm not familiar enough with the over class and problem to provide much useful feedback.

**Comment by rnugent3** [comment]
Last night, I dreamed up an alternative solution that would not require sorting. I'm going to make some changes tomorrow or over the weekend, and then let's talk.

**Comment by rnugent3** [comment]
the compute appears to be pretty fast - in 40 seconds we can run 2 million iterations.

I need to add back in the histogram for plotting assurance of aep

**Comment by rnugent3** [comment]
tests failing on remote but passing locally are my favorite tests

**Comment by rnugent3** [comment]
the tests test more than the code............

File: HEC.FDA.Model/structures/OccupancyType.cs [hunk 2]
```diff
@@ -201,6 +201,13 @@ internal OccupancyTypeBuilder(OccupancyType occupancyType)
             public OccupancyType build()
             {
                 _occupancyType.Validate();
+                //unregister?
+                //unregister each of the objects that got built within the builder pattern 
+                //reflect on an instance of this, ask for any field that is a validation data error logger thingy
+                //on that thing, if not null, unregister 
+
+                //need something like IUnregisterFromMessageHub
+                //in computational method, at the very beginning, we call Validate()
```
Comment (rnugent3) [inline HEC.FDA.Model/structures/OccupancyType.cs:210]: Kill commented code

File: HEC.FDA.Statistics/Distributions/IDistributionFactory.cs [hunk 3]
```diff
@@ -38,13 +38,19 @@ public static string PrintParamaterizationRequirements(IDistributionEnum type)
...
             }
         }
-
+        /// <summary>
+        /// we have to keep this because of IDistribution
+        /// </summary>
+        /// <param name="sample"></param>
+        /// <param name="nBins"></param>
+        /// <returns></returns>
         internal static Statistics.Histograms.Histogram Fit(IEnumerable<double> sample, int nBins)
```
Comment (rnugent3) [inline HEC.FDA.Statistics/Distributions/IDistributionFactory.cs:47]: IDistribution has a bunch of baggage. That interface needs broken up into multiple interfaces so classes don't have to implement bogus methods.

Comment (rnugent3) [reply HEC.FDA.Statistics/Distributions/IDistributionFactory.cs:47]: Also, Delete IDistribution Factory and replace it with the proper constructors all over the place. It's only used in tests .

File: HEC.FDA.Statistics/Histograms/ThreadsafeInlineHistogram.cs [hunk 4]
```diff
@@ -195,6 +195,7 @@ public ThreadsafeInlineHistogram()
                 AddObservationToHistogram(0,i);
             }
         }
+
```
Comment (rnugent3) [inline HEC.FDA.Statistics/Histograms/ThreadsafeInlineHistogram.cs:198]: Delete ThreadsafeInlineHistograms

**Pull request merged**